homeassistant/components/media_player/kodi.py

```diff
@@ -593,6 +654,141 @@ def async_play_media(self, media_type, media_id, **kwargs):
         if media_type == "CHANNEL":
             return self.server.Player.Open(
                 {"item": {"channelid": int(media_id)}})
+        elif media_type == "PLAYLIST":
+            return self.server.Player.Open(
+                {"item": {"playlistid": int(media_id)}})
         else:
             return self.server.Player.Open(
                 {"item": {"file": str(media_id)}})
+
+    @asyncio.coroutine
+    def async_set_shuffle(self, shuffle_on):
+        """Set shuffle mode, for the first player."""
+        players = yield from self._get_players()
```

Comment:
There is no need to fetch the players, they are cached as `self._players` when the state is updated.